### DISCUSSION TITLE
how to save an image to the database using pocketbase hooks files?

### DISCUSSION THREAD

**Discussion by AxelJPG** [pocketbase/pocketbase#3653] [Q&A]
I'm trying to save an image that a form sends but I don't know the correct way to do it, try "addFiles
"but it does not save the information

**Comment by ganigeorgiev** [comment]
I don't understand what do you mean by _"try addFiles but it does not save the information"_. What information?

Could you provide a code sample with what you've tried and more details about your collections structure (_you can export your collections configuration as json from the Admin UI > Settings > Export collections_)?

---

In general, for creating a record programmatically with data validations you can check the example in https://pocketbase.io/docs/js-records/#create-new-record-with-data-validations.

Here is another example, showing `form.loadRequest` usage and manually loading multipart/form-data files via `addFiles` (_I haven't tried it; see also https://github.com/pocketbase/pocketbase/discussions/2489#discussioncomment-7355379_):

```js
// cusotm /upload route
routerAdd("POST", "/upload", (c) => {
    const collection = $app.dao().findCollectionByNameOrId("example")

    const record = new Record(collection)

    const form = new RecordUpsertForm($app, record)

    // load the entire request
    // (note that submitted field names must match with the ones from the collection)
    form.loadRequest(c.request(), "")

    // ----------------------------------
    // AND/OR manually populate the form data
    // ----------------------------------

    // regular fields
    // (to read request body fields - https://pocketbase.io/docs/js-routing/#reading-request-body)
    form.loadData({
        "someField": 123,
    })

    // multipart/form-data uploaded files
    const [_, multipartHeader] = c.request().formFile("yourFileField")

    form.addFiles("yourFileField", $filesystem.fileFromMultipart(multipartHeader))

    try {
        form.submit()
    } catch (err) {
        throw new BadRequestError("failed to create record", err)
    }

    return c.json(200, record)
})
```

**Comment by AxelJPG** [comment]
I'm sorry I didn't explain myself clearly, I'm new to this.
The database I have has a collection with an image field, I am trying to save the image that comes from a form using "Extend with JavaScript". I created a new route to achieve this but the problem I have is that I don't know how to do it 🥲

This is the form

```typescript
import PocketBase from 'pocketbase'
import { useForm, SubmitHandler } from 'react-hook-form'

const pb = new PocketBase('http://localhost:8090')
pb.autoCancellation(false)

type Inputs = {
  nombre: string
  logo: File
}

const NewStore = async (values: Inputs) => {
  await pb.send('/store', {
    method: 'POST',
    body: values
  })
}

export const App = () => {
  const { register, handleSubmit } = useForm<Inputs>()
  const onSubmit: SubmitHandler<Inputs> = (data) => {
    NewStore(data)
  }

  return (
    <form onSubmit={handleSubmit(onSubmit)} encType='multipart/form-data'>
      <input required defaultValue='test-store' {...register('nombre')} />
      <input required type='file' {...register('logo')} />
      <input type='submit' />
    </form>
  )
}

``` 
And this is my hook
I used the way you recommended but I'm not sure if I implemented it correctly

```javascript
routerAdd('POST', '/store', (c) => {
  const collection = $app.dao().findCollectionByNameOrId('example')
  const record = new Record(collection)
  const form = new RecordUpsertForm($app, record)
  form.loadRequest(c.request(), '')

  form.loadData({
    nombre: body?.nombre
  })

  const [_, multipartHeader] = c.request().formFile(body?.logo)

  form.addFiles('yourFileField', $filesystem.fileFromMultipart(multipartHeader))

  try {
    form.submit()
  } catch (err) {
    throw new BadRequestError('failed to create record', err)
  }

  return c.json(200, record)
})

``` 

Sorry for the inconveniences

**Reply by ganigeorgiev** [reply]
I don't understand the above code.

Also if you don't have custom persistence logic why do you need a new `pb_hooks` route and not use directly the existing collection APIs?

Or in other words, why don't send from your react form a direct request like:

```js
await pb.collection("example").create(data)
```
_(see also https://pocketbase.io/docs/files-handling/#uploading-files)_

I don't think I'll be able to help you without information about your use case or some basic overview of what you are trying to do.
If the above doesn't help, please elaborate and provide an example with desired outcome (pseudo-code is also ok).

**Comment by AxelJPG** [comment]
The software is for a store that manages small businesses, create a new route pb_hook because I need to validate if the code associated with the business already exists.
I did it this way so as not to overload the client application with logic.

I am very grateful for your help.

**Reply by ganigeorgiev** [reply]
I'm sorry, I don't know how to help you and I really hate this tedious back-and-forth.

Please try to be more specific.

I know that it takes time to explain problems, especially if you are not a native speaker, but also please understand that I have literally no context about your project, nor what _"code associated with the business already exists"_ is supposed to mean. What is even "code" in the first place? Is this `body.nombre` in the above?

Do you really need a custom route for your validations? Did you try using the [`onRecordBeforeCreate` hook](https://pocketbase.io/docs/js-event-hooks/#onrecordbeforecreaterequest) instead?

Alternatively, if you just want a unique field validation, you can try creating a unique index for your field from the "Unique constraints and indexes" collection settings in the Admin UI.
![unique_index_screenshot](https://i.imgur.com/jasiCLM.png)

But again, I'm not ChatGPT and I can't hallucinate an answer for you. I need to understand the problem to at least propose a suggestion or workaround, otherwise we are just going in circles.

**Comment by AxelJPG** [comment]
Thank you very much, I will use the recommendation you provided. 🙇🏻
